src/Orleans/Utils/CachedReadConcurrentDictionary.cs

```diff
@@ -0,0 +1,203 @@
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Orleans.Utilities
+{
+    /// <summary>
+    /// A thread-safe dictionary for read-heavy workloads.
+    /// </summary>
+    /// <typeparam name="TKey">The key type.</typeparam>
+    /// <typeparam name="TValue">The value type.</typeparam>
+    internal class CachedReadConcurrentDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+    {
+        /// <summary>
+        /// The number of cache misses which are tolerated before the cache is regenerated.
+        /// </summary>
+        private const int CacheMissesBeforeCaching = 10;
+        private readonly ConcurrentDictionary<TKey, TValue> dictionary;
+        private readonly IEqualityComparer<TKey> comparer;
+
+        /// <summary>
+        /// Approximate number of reads which did not hit the cache since it was last invalidated.
+        /// This is used as a heuristic that the dictionary is not being modified frequently with respect to the read volume.
+        /// </summary>
+        private int cacheMissReads;
+
+        /// <summary>
+        /// Cached version of <see cref="dictionary"/>.
+        /// </summary>
+        private volatile Dictionary<TKey, TValue> readCache;
```

Comment:
Is there a real need for `volatile` here? Just want to reduce the complexity if it's not.
Because it looks like if you remove volatile, you still can read your writes, and if another thread wants to access it, it probably doesn't matter that the cache was stale (otherwise more attention needs to be paid to when you invalidate the cache, which in this case it looks like is after you modify it, so there are still chances that the cache is stale when read from another thread, even with the volatile).
Or am I missing some edge case?